In [1]:
#Prints **all** console output, not just last item in cell 
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [2]:
from boilerplate import *
# from probdist import *

In [3]:
import os

In [4]:
# see https://github.com/scjs/buckeye 
# and esp. https://nbviewer.jupyter.org/github/scjs/buckeye/blob/master/Quickstart.ipynb
import buckeye as b

The goals here are
 1. generally to expose the structure of the data (as viewed via the `buckeye` package).
 2. specifically to collect all the phonemes and phones in the data and see if there are any non-trivial surprises lurking in the inventories or transcriptions.
 3. ..and to also collect all of the orthographic wordforms in Buckeye.

# Buckeye data: import and look around

In [5]:
%pwd

'/mnt/cube/home/AD/emeinhar/wr'

In [6]:
%cd ..

/mnt/cube/home/AD/emeinhar


In [7]:
%ls

anaconda3/  buckeye-zips/  Downloads/  GitHub/  PID/
buckeye/    cmu-ipa/       evo/        lexd/    wr/


In [8]:
%cd buckeye-zips/

/mnt/cube/home/AD/emeinhar/buckeye-zips


In [9]:
%ls

s01.zip  s05.zip  s10.zip  s17.zip  s24.zip  s29.zip  s35.zip  s39.zip
s02.zip  s06.zip  s12.zip  s18.zip  s26.zip  s30.zip  s36.zip  s40.zip
s03.zip  s07.zip  s15.zip  s19.zip  s27.zip  s31.zip  s37.zip
s04.zip  s09.zip  s16.zip  s21.zip  s28.zip  s34.zip  s38.zip


In [10]:
files = os.listdir(); files

['s05.zip',
 's02.zip',
 's39.zip',
 's37.zip',
 's30.zip',
 's21.zip',
 's26.zip',
 's28.zip',
 's31.zip',
 's36.zip',
 's38.zip',
 's03.zip',
 's04.zip',
 's15.zip',
 's12.zip',
 's29.zip',
 's27.zip',
 's16.zip',
 's18.zip',
 's24.zip',
 's35.zip',
 's40.zip',
 's09.zip',
 's07.zip',
 's19.zip',
 's17.zip',
 's10.zip',
 's01.zip',
 's06.zip',
 's34.zip']

In [11]:
speakers = [b.Speaker.from_zip(file) for file in files]

In [13]:
tracks = list(union([set(s.tracks) for s in speakers]))
len(tracks)

193

In [14]:
t = tracks[0]; type(t)
t

buckeye.buckeye.Track

Track("s2602b")

In [18]:
Ws = t.words[:10]
Ws

[Pause('{B_TRANS}', 0.0, 0.241961),
 Word('kids', 0.241961, 0.979501, ['k', 'ih', 'd', 'z'], ['k', 'ih', 'd', 's'], 'NNS'),
 Pause('<SIL>', 0.979501, 2.189889),
 Word('because', 2.189889, 2.383308, ['b', 'ih', 'k', 'ah', 'z'], ['k', 'ah', 'z'], 'IN'),
 Word("they're", 2.383308, 2.493646, ['dh', 'ey', 'er'], ['eh', 'r'], 'PRP_VBP'),
 Word('only', 2.493646, 2.748, ['ow', 'n', 'l', 'iy'], ['ow', 'n', 'l', 'iy'], 'RB'),
 Word('little', 2.748, 3.031915, ['l', 'ih', 't', 'el'], ['l', 'ih', 't', 'el'], 'RB'),
 Word('once', 3.031915, 3.847954, ['w', 'ah', 'n', 's'], ['w', 'ah', 'n', 's'], 'RB'),
 Pause('<SIL>', 3.847954, 4.250971),
 Word("that's", 4.250971, 4.508236, ['dh', 'ae', 't', 's'], ['th', 'ah', 't', 's'], 'DT_VBZ')]

In [25]:
w0 = Ws[0]
w0

Pause('{B_TRANS}', 0.0, 0.241961)

In [28]:
w0.entry
type(w0)

'{B_TRANS}'

buckeye.containers.Pause

In [20]:
w1 = Ws[1]
w1

Word('kids', 0.241961, 0.979501, ['k', 'ih', 'd', 'z'], ['k', 'ih', 'd', 's'], 'NNS')

In [24]:
w1.phonemic
w1.phonetic

['k', 'ih', 'd', 'z']

['k', 'ih', 'd', 's']

In [44]:
def getPhonemes(word):
    return word.phonemic

def getPhones(word):
    return word.phonetic

In [30]:
b.containers.Word
type(w1) == b.containers.Word

buckeye.containers.Word

True

In [42]:
def isWord(container):
    return type(container) == b.containers.Word

def getWords(track):
    return [w for w in track.words if isWord(w)]
#     return filter(isWord, track.words)

In [37]:
Ws_0 = list(getWords(t))
len(Ws_0)

846

In [38]:
Ws_0[:10]

[Word('kids', 0.241961, 0.979501, ['k', 'ih', 'd', 'z'], ['k', 'ih', 'd', 's'], 'NNS'),
 Word('because', 2.189889, 2.383308, ['b', 'ih', 'k', 'ah', 'z'], ['k', 'ah', 'z'], 'IN'),
 Word("they're", 2.383308, 2.493646, ['dh', 'ey', 'er'], ['eh', 'r'], 'PRP_VBP'),
 Word('only', 2.493646, 2.748, ['ow', 'n', 'l', 'iy'], ['ow', 'n', 'l', 'iy'], 'RB'),
 Word('little', 2.748, 3.031915, ['l', 'ih', 't', 'el'], ['l', 'ih', 't', 'el'], 'RB'),
 Word('once', 3.031915, 3.847954, ['w', 'ah', 'n', 's'], ['w', 'ah', 'n', 's'], 'RB'),
 Word("that's", 4.250971, 4.508236, ['dh', 'ae', 't', 's'], ['th', 'ah', 't', 's'], 'DT_VBZ'),
 Word('what', 4.508236, 4.591645, ['w', 'ah', 't'], ['w', 'ah', 'dx'], 'WP'),
 Word('i', 4.591645, 4.667259, ['ay'], ['ay'], 'PRP'),
 Word('tell', 4.667259, 4.863387, ['t', 'eh', 'l'], ['t', 'eh', 'l'], 'VBP')]

In [43]:
words = list(union([set(getWords(t)) for t in tracks]))
len(words)

216062

In [45]:
phonemes = set(union([set(getPhonemes(w)) for w in words]))
phones = set(union([set(getPhones(w)) for w in words]))

In [46]:
len(phonemes)
len(phones)

46

79

In [52]:
phonemes

{'aa',
 'ae',
 'ah',
 'ao',
 'aw',
 'ay',
 'b',
 'ch',
 'd',
 'dh',
 'dx',
 'eh',
 'el',
 'em',
 'en',
 'er',
 'ey',
 'f',
 'g',
 'h',
 'hh',
 'ih',
 'iy',
 'jh',
 'k',
 'l',
 'm',
 'n',
 'ng',
 'nx',
 'ow',
 'oy',
 'p',
 'r',
 's',
 'sh',
 't',
 'th',
 'tq',
 'uh',
 'uw',
 'v',
 'w',
 'y',
 'z',
 'zh'}

In [48]:
phones

{'IVER',
 'LAUGH',
 'NOISE',
 'SIL',
 'UNKNOWN',
 'VOCNOISE',
 'a',
 'aa',
 'aan',
 'ae',
 'aen',
 'ah',
 'ahn',
 'an',
 'ao',
 'aon',
 'aw',
 'awn',
 'ay',
 'ayn',
 'b',
 'ch',
 'd',
 'dh',
 'dx',
 'eh',
 'ehn',
 'el',
 'em',
 'en',
 'eng',
 'er',
 'ern',
 'ey',
 'eyn',
 'f',
 'g',
 'h',
 'hh',
 'hhn',
 'i',
 'id',
 'ih',
 'ihn',
 'iy',
 'iyih',
 'iyn',
 'jh',
 'k',
 'l',
 'm',
 'n',
 'ng',
 'nx',
 'ow',
 'own',
 'oy',
 'oyn',
 'p',
 'q',
 'r',
 's',
 'sh',
 't',
 'th',
 'tq',
 'uh',
 'uhn',
 'uw',
 'uwix',
 'uwn',
 'v',
 'w',
 'x',
 'y',
 'z',
 'zh',
 '{B_TRANS}',
 '{E_TRANS}'}

There are some remarkable objects in the phone inventory:

In [49]:
badPhones = {
 'IVER',
 'LAUGH',
 'NOISE',
 'SIL',
 'UNKNOWN',
 'VOCNOISE',
 '{B_TRANS}',
 '{E_TRANS}'}
badPhones

{'IVER',
 'LAUGH',
 'NOISE',
 'SIL',
 'UNKNOWN',
 'VOCNOISE',
 '{B_TRANS}',
 '{E_TRANS}'}

In [50]:
def hasBadPhones(word):
    phones = getPhones(word)
    return any([p in phones for p in badPhones])

In [54]:
badPhonesWords = [w for w in words if hasBadPhones(w)]
len(badPhonesWords)
len(words)
len(badPhonesWords) / len(words)
badPhonesWords[:10]

109

216062

0.0005044848238005757

[Word('these', 304.848233, 304.976757, ['dh', 'iy', 'z'], ['dh', 'ey', 'UNKNOWN'], 'DT'),
 Word('like', 407.7053, 407.998881, ['l', 'ay', 'k'], ['SIL', 'l', 'ay', 'k'], 'IN'),
 Word("i'm", 77.13, 78.095, ['ay', 'm'], ['IVER', 'ah', 'm'], 'PRP_VBP'),
 Word('like', 468.190022, 468.622004, ['l', 'ay', 'k'], ['VOCNOISE', 'l', 'ay', 'k'], 'IN'),
 Word('i', 78.30982, 78.482, ['ay'], ['SIL', 'ay'], 'PRP'),
 Word('other', 582.264738, 582.497257, ['ah', 'dh', 'er'], ['VOCNOISE', 'ah', 'dh', 'ah'], 'JJ'),
 Word('yknow', 540.89, 541.264, ['y', 'uw', 'n', 'ow'], ['SIL', 'y', 'ih', 'n', 'ao'], 'PRP_VBP'),
 Word('drafting', 205.590042, 206.311723, ['d', 'r', 'ae', 'f', 't', 'ih', 'ng'], ['d', 'r', 'ae', 'UNKNOWN', 't', 'ih', 'ng'], 'VBG'),
 Word('and', 549.547501, 549.639, ['ae', 'n', 'd'], ['SIL', 'eh', 'nx'], 'CC'),
 Word("it's", 593.185488, 593.63, ['ih', 't', 's'], ['SIL', 'ih', 't', 's'], 'PRP_VBZ')]

Only $\frac{109}{216062}$ word tokens in the corpus have one of these unusual elements. I think those words can eventually be excluded without loss of generality.

In [56]:
goodPhones = phones - badPhones
len(phones)
len(goodPhones)
len(badPhones)

79

71

8

# Collecting orthographic wordforms

In [62]:
def getOrthography(word):
    return word.orthography

In [63]:
w1
w1.orthography
getOrthography(w1)

Word('kids', 0.241961, 0.979501, ['k', 'ih', 'd', 'z'], ['k', 'ih', 'd', 's'], 'NNS')

'kids'

'kids'

In [65]:
orthographic_wordforms = list(map(getOrthography,
                                  words))

In [70]:
orthographic_wordform_counter = Counter(orthographic_wordforms)
sum(orthographic_wordform_counter.values())
orthographic_wordform_counter.most_common(n=100)

216062

[('i', 9418),
 ('and', 8227),
 ('the', 6204),
 ('to', 4889),
 ('a', 4549),
 ('that', 4275),
 ('it', 3520),
 ('like', 3504),
 ('you', 3475),
 ('of', 3091),
 ('yknow', 2847),
 ('they', 2835),
 ('in', 2580),
 ('uh', 2556),
 ("it's", 2292),
 ('but', 2248),
 ('was', 2227),
 ('so', 2035),
 ('yeah', 1994),
 ('just', 1979),
 ("don't", 1721),
 ('um', 1692),
 ('have', 1623),
 ('know', 1484),
 ('is', 1466),
 ('my', 1378),
 ('he', 1375),
 ('think', 1366),
 ('for', 1329),
 ('we', 1326),
 ('do', 1298),
 ('mean', 1214),
 ('or', 1178),
 ('there', 1173),
 ('be', 1159),
 ('on', 1149),
 ('not', 1148),
 ('with', 1136),
 ('because', 1122),
 ('really', 1107),
 ("that's", 1090),
 ('what', 1014),
 ('well', 1000),
 ('people', 985),
 ('if', 981),
 ("i'm", 940),
 ('at', 910),
 ('all', 908),
 ('this', 901),
 ('get', 877),
 ('go', 859),
 ('are', 850),
 ('about', 848),
 ('when', 835),
 ('them', 819),
 ('out', 795),
 ('she', 785),
 ('me', 760),
 ('then', 730),
 ('had', 730),
 ('as', 716),
 ('one', 715),
 ('would', 7

In [71]:
unique_wordforms = set(orthographic_wordforms)
len(unique_wordforms)

7999

In [79]:
t.words

[Pause('{B_TRANS}', 0.0, 0.241961),
 Word('kids', 0.241961, 0.979501, ['k', 'ih', 'd', 'z'], ['k', 'ih', 'd', 's'], 'NNS'),
 Pause('<SIL>', 0.979501, 2.189889),
 Word('because', 2.189889, 2.383308, ['b', 'ih', 'k', 'ah', 'z'], ['k', 'ah', 'z'], 'IN'),
 Word("they're", 2.383308, 2.493646, ['dh', 'ey', 'er'], ['eh', 'r'], 'PRP_VBP'),
 Word('only', 2.493646, 2.748, ['ow', 'n', 'l', 'iy'], ['ow', 'n', 'l', 'iy'], 'RB'),
 Word('little', 2.748, 3.031915, ['l', 'ih', 't', 'el'], ['l', 'ih', 't', 'el'], 'RB'),
 Word('once', 3.031915, 3.847954, ['w', 'ah', 'n', 's'], ['w', 'ah', 'n', 's'], 'RB'),
 Pause('<SIL>', 3.847954, 4.250971),
 Word("that's", 4.250971, 4.508236, ['dh', 'ae', 't', 's'], ['th', 'ah', 't', 's'], 'DT_VBZ'),
 Word('what', 4.508236, 4.591645, ['w', 'ah', 't'], ['w', 'ah', 'dx'], 'WP'),
 Word('i', 4.591645, 4.667259, ['ay'], ['ay'], 'PRP'),
 Word('tell', 4.667259, 4.863387, ['t', 'eh', 'l'], ['t', 'eh', 'l'], 'VBP'),
 Word('myself', 4.863387, 5.343045, ['m', 'ay', 's', 'eh', 'l'

limit_output extension: Maximum message size of 20000 exceeded with 76465 characters

In [82]:
w1.phones

[Phone('k', 0.241961, 0.394966),
 Phone('ih', 0.394966, 0.673357),
 Phone('d', 0.673357, 0.801466),
 Phone('s', 0.801466, 0.978397)]